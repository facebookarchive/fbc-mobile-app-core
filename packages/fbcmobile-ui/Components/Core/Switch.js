/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow strict-local
 * @format
 */

'use strict';

import * as React from 'react';
import {
  Animated,
  Easing,
  StyleSheet,
  TouchableWithoutFeedback,
  View,
} from 'react-native';
import {Colors} from '@fbcmobile/ui/Theme';
import {Icon} from 'react-native-material-ui';
import {useLayoutEffect, useRef, useState} from 'react';

type Props = {
  +checked: boolean,
  +onPress?: () => void,
};

const Switch = ({checked, onPress}: Props) => {
  const [anim] = useState(new Animated.Value(checked === true ? 1 : 0));
  const animatedContainerStyle = useRef(AnimatedStyles.containerStyle(anim))
    .current;
  const animatedLeverStyle = useRef(AnimatedStyles.leverStyle(anim)).current;

  useLayoutEffect(() => {
    Animated.timing(
      anim,
      AnimatedStyles.animate(checked === true ? 1 : 0),
    ).start();
  }, [anim, checked]);

  return (
    <TouchableWithoutFeedback onPress={onPress}>
      <View>
        <Animated.View
          style={[styles.containerStyle, animatedContainerStyle]}
        />
        <Animated.View style={[styles.leverStyle, animatedLeverStyle]}>
          {checked && (
            <Icon
              name="check"
              iconSet="MaterialIcons"
              color={Colors.Blue}
              size={12}
            />
          )}
        </Animated.View>
      </View>
    </TouchableWithoutFeedback>
  );
};

const outputRange: Array<string> = [Colors.DisabledText, Colors.Blue];
const leverOutputRange: Array<number> = [0, 14];

const AnimatedStyles = {
  containerStyle(anim) {
    return {
      backgroundColor: anim.interpolate({
        inputRange: [0, 1],
        outputRange: outputRange,
        extrapolate: 'clamp',
      }),
    };
  },
  leverStyle(anim) {
    return {
      transform: [
        {
          translateX: anim.interpolate({
            inputRange: [0, 1],
            outputRange: leverOutputRange,
            extrapolate: 'clamp',
          }),
        },
      ],
    };
  },
  animate(toValue) {
    return {
      toValue,
      duration: 150,
      easing: Easing.linear,
      useNativeDriver: false,
    };
  },
};

const styles = StyleSheet.create({
  leverStyle: {
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
    backgroundColor: Colors.White,
    height: 20,
    width: 20,
    borderRadius: 20,
    position: 'absolute',
    top: 1,
    left: 1,
  },
  containerStyle: {
    width: 36,
    height: 22,
    borderRadius: 11,
    padding: 1,
    backgroundColor: Colors.DisabledText,
  },
});

export default Switch;
