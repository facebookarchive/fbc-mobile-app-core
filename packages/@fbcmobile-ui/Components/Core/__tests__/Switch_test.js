/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow strict-local
 * @format
 */

'use strict';

import React from 'react';
import Switch from '../Switch';
import renderer from 'react-test-renderer';

describe('Switch Tests', () => {
  test('checked switch renders correctly', () => {
    const tree = renderer.create(<Switch checked={true} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
  test('un-checked switch renders correctly', () => {
    const tree = renderer.create(<Switch checked={false} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
