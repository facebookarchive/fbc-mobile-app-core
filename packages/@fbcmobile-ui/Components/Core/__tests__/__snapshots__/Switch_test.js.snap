// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch Tests checked switch renders correctly 1`] = `
<View
  accessible={true}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
>
  <View
    style={
      Object {
        "backgroundColor": "rgba(24, 119, 242, 1)",
        "borderRadius": 11,
        "height": 22,
        "padding": 1,
        "width": 36,
      }
    }
  />
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "white",
        "borderRadius": 20,
        "display": "flex",
        "height": 20,
        "justifyContent": "center",
        "left": 1,
        "position": "absolute",
        "top": 1,
        "transform": Array [
          Object {
            "translateX": 14,
          },
        ],
        "width": 20,
      }
    }
  >
    <Text
      allowFontScaling={false}
      style={
        Array [
          Object {
            "color": "#1877F2",
            "fontSize": 12,
          },
          null,
          Object {
            "fontFamily": "Material Icons",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          Object {},
        ]
      }
    >
      î—Š
    </Text>
  </View>
</View>
`;

exports[`Switch Tests un-checked switch renders correctly 1`] = `
<View
  accessible={true}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
>
  <View
    style={
      Object {
        "backgroundColor": "rgba(188, 192, 196, 1)",
        "borderRadius": 11,
        "height": 22,
        "padding": 1,
        "width": 36,
      }
    }
  />
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "white",
        "borderRadius": 20,
        "display": "flex",
        "height": 20,
        "justifyContent": "center",
        "left": 1,
        "position": "absolute",
        "top": 1,
        "transform": Array [
          Object {
            "translateX": 0,
          },
        ],
        "width": 20,
      }
    }
  />
</View>
`;
